loadtest.go

```diff
@@ -89,14 +91,163 @@ func (l *loadTest) runUser() {
 	l.log.Debug("user stopped")
 }
 
+func registerPushDevice(config *config.Config, deviceID, outputChannel string, rest *ably.REST, log log15.Logger) error {
+	type recipient struct {
+		TransportType string `json:"transportType"`
+		Channel       string `json:"channel"`
+		AblyKey       string `json:"ablyKey"`
+		AblyUrl       string `json:"ablyUrl"`
+	}
+	type push struct {
+		Recipient recipient `json:"recipient"`
+	}
+	type input struct {
+		Id         string `json:"id"`
+		Platform   string `json:"platform"`
+		FormFactor string `json:"formFactor"`
+		Push       push   `json:"push"`
+	}
+	regInput := &input{
+		Id:         deviceID,
+		Platform:   "browser",
+		FormFactor: "other",
+		Push: push{
+			Recipient: recipient{
+				TransportType: "ablyChannel",
+				Channel:       outputChannel,
+				AblyKey:       config.Ably.APIKey,
+				AblyUrl:       config.Subscriber.PushDevice.URL,
+			},
+		},
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/deviceRegistrations",
+		nil,
+		regInput,
+		nil)
+	if err != nil {
+		log.Debug("error registering a push device", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error registering a push device", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
+func subscribePushDevice(config *config.Config, deviceID, channel string, rest *ably.REST, log log15.Logger) error {
+	type input struct {
+		Channel  string `json:"channel"`
+		DeviceId string `json:"deviceId"`
+	}
+	subInput := &input{
+		Channel:  channel,
+		DeviceId: deviceID,
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/channelSubscriptions",
+		nil,
+		subInput,
+		nil)
+	if err != nil {
+		log.Debug("error subscribing to a channel", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error subscribing to a channel", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
 // runSubscriber runs a subscriber task which renders the channel names using
 // the given user number and subscribes to each of them.
 func (l *loadTest) runSubscriber(ctx context.Context, client Client, userNum int64) error {
 	channels := renderChannels(l.subscriberChannels, userNum)
 
+	errG, ctx := errgroup.WithContext(ctx)
+	if l.w.Conf().Subscriber.PushDevice.Enabled {
+		l.log.Debug("creating push device")
+
+		if l.w.Conf().Subscriber.PushDevice.MetachannelEnabled {
+			errG.Go(func() error {
+				for {
+					l.log.Debug("subscribing to metachannel")
+					err := client.Subscribe(ctx, "[meta]log:push", func(data []byte) {
+						l.log.Info("push metachannel:", "message", string(data))
```

Comment:
If this is logging on a per message basis, let's make it a debug line, otherwise it will flood the logs in a load test.

Also, what is the format of `data`? It would be nice to record a failure if the data indicates there was an error.

Reply:
> If this is logging on a per message basis, let's make it a debug line, otherwise it will flood the logs in a load test.

I though it would be a bit odd to activate that option and not see anything because we disabled the debug log level, but I understand the issue.

> Also, what is the format of data? It would be nice to record a failure if the data indicates there was an error.

I had a quick look, but it seems to use the logging format: `map[action:push.channelSubscription.update error:<nil> message:subscription updated meta:map[appId:xxxxx channelSubscription:map[appId:xxxxx channel:pushenabled:fanout deviceId:device-e111a7fa] site:eu-west-2-A tag:push] severity:info tag:push timestamp:1.61968664973e+12]` that does seem non-trivial, so I considered it was not worth it.

Reply:
I'm not sure why the push worker isn't marshalling into JSON. :thinking:

Reply:
Ah ok, we should fix that, it's not very useful in that format, so for now we'll just log it then 👍